File: langs/clojure/tetris-polylith/components/move/src/tetrisanalyzer/move/placement.clj
(ns tetrisanalyzer.move.placement
  (:require [tetrisanalyzer.move.move :as move]
            [tetrisanalyzer.move.visit :as visit]
            [tetrisanalyzer.board.interface :as board]
            [tetrisanalyzer.piece.interface :as piece]))

(defn placements
  [board x y p rotation bitmask valid-moves visited-moves rotation-fn shapes]
  (loop [next-moves (list [x y rotation])
         placements []
         valid-moves valid-moves
         visited-moves visited-moves]
    (if-let [[x y rotation] (first next-moves)]
      (let [next-moves (rest next-moves)]
        (if (visit/visited? visited-moves x y rotation)
          (recur next-moves placements valid-moves visited-moves)
          (let [[down placement] (move/down board x y p rotation bitmask shapes)
                moves (keep #(% board x y p rotation bitmask shapes)
                            [move/left
                             move/right
                             rotation-fn
                             (constantly down)])]
            (recur (into next-moves moves)
                   (concat placements placement)
                   (conj valid-moves [x y rotation])
                   (visit/visit visited-moves x y rotation)))))
      placements)))

(defn all-placements [board p start-x kick? shapes]
  (let [bitmask (piece/rotation-bitmask shapes p)
        visited-moves (board/empty-board board)
        rotation-fn (move/rotation-fn kick?)]
    (if (move/valid-move? board start-x 0 p 0 shapes)
      (placements board start-x 0 p 0 bitmask [] visited-moves rotation-fn shapes)
      [])))

File: langs/python/tetris-polylith-uv/components/tetrisanalyzer/piece/placement/placement.py
from collections import deque

from tetrisanalyzer import board as board_ifc, piece
from . import move, visit


def _placements(board, x, y, p, rotation, bitmask, valid_moves, visited_moves, rotation_move_fn, shapes):
    next_moves = deque([[x, y, rotation]])
    placements = []

    while next_moves:
        x, y, rotation = next_moves.popleft()

        if visit.is_visited(visited_moves, x, y, rotation):
            continue

        down_move, placement = move.down(board, x, y, p, rotation, bitmask, shapes)

        moves = [
            move.left(board, x, y, p, rotation, bitmask, shapes),
            move.right(board, x, y, p, rotation, bitmask, shapes),
            rotation_move_fn(board, x, y, p, rotation, bitmask, shapes),
            down_move,
        ]
        moves = [move for move in moves if move is not None]

        next_moves.extend(moves)

        if placement is not None:
            placements.extend(placement)

        valid_moves.append([x, y, rotation])
        visit.visit(visited_moves, x, y, rotation)

    return placements


def placements(board, p, start_x, kick, shapes):
    y = 0
    rotation = 0
    bitmask = piece.bitmask.rotation_bitmask(shapes, p)
    visited_moves = board_ifc.empty_board_as(board)
    rotation_move_fn = move.rotation_fn(kick)

    if not move.is_valid_move(board, start_x, y, p, rotation, shapes):
        return []

    return _placements(board, start_x, y, p, rotation, bitmask, [], visited_moves, rotation_move_fn, shapes)
